cgp/genome.py

```diff
@@ -267,6 +267,172 @@ def randomize(self, rng: np.random.RandomState) -> None:
+
+        # determine for each node where it receives input from
+        dependencies = self._determine_dependencies()
+
+        # initialize addable set to internal nodes with empty set
+        addable_nodes = set(
+            idx
+            for (idx, item) in enumerate(dependencies)
+            if not item and self._is_hidden_region(idx)
+        )
+        node_idx = self._n_inputs  # First position to be placed is after inputs
+        used_idx_list = []
+
+        while len(addable_nodes) > 0:
+            # pick one of the addable nodes
+            current_idx = rng.choice(list(addable_nodes))
+
+            # take this nodes dna sequence
+            node_dna = self._dna[
+                current_idx * length_per_region : (current_idx + 1) * length_per_region
+            ]
+            # and insert it at the working location of the copy
+            dna_for_reorder[
+                node_idx * length_per_region : (node_idx + 1) * length_per_region
+            ] = node_dna
+
+            # remove the current idx from the dependencies set
+            for current_dependencies in dependencies:
+                current_dependencies.discard(current_idx)
+
+            # update set of addable nodes
+            addable_nodes = set(
+                idx
+                for (idx, item) in enumerate(dependencies)
+                if not item and self._is_hidden_region(idx)
+            )
+            # remove current_idx and all previous used idx's
+            # from addable_nodes and increment node_idx counter
+            used_idx_list.append(current_idx)
+            addable_nodes = addable_nodes.difference(used_idx_list)
+            node_idx += 1
+
+        # change the gene values such that they fit the new location of the genes
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                try:
+                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
+                except ValueError:  # means the input is from an input node -> do nothing
+                    pass
+            dna_for_reorder[gene_idx] = gene_value
+
+        # change the invalid gene values (coming from unused arities)
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) and gene_value > region_idx:
+                # replace value that is too large by smaller value
+                permissible_values = self.determine_permissible_values_per_gene(gene_idx)
+                gene_value = rng.choice(permissible_values)
+                dna_for_reorder[gene_idx] = gene_value
+
+        self.dna = dna_for_reorder
+
+    def _determine_dependencies(self) -> List[set]:
+
+        dependencies = []
```

Comment:
since you iterate over the regions in order you can use a list here and the index it with `input_region_idx` below. just to be a bit more transparent, maybe use a dict here, mapping region indices to their dependencies?